himbaechel/uarch/gatemate/route_mult.cc

```diff
@@ -0,0 +1,826 @@
+/*
+ *  nextpnr -- Next Generation Place and Route
+ *
+ *  Copyright (C) 2024  The Project Peppercorn Authors.
+ *
+ *  Permission to use, copy, modify, and/or distribute this software for any
+ *  purpose with or without fee is hereby granted, provided that the above
+ *  copyright notice and this permission notice appear in all copies.
+ *
+ *  THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ *  WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ *  MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ *  ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ *  WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN AN
+ *  ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT OF
+ *  OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ *
+ */
+
+#include "gatemate.h"
+#include "log.h"
+#include "nextpnr_assertions.h"
+#include "nextpnr_base_types.h"
+#include "nextpnr_namespaces.h"
+
+#define HIMBAECHEL_CONSTIDS "uarch/gatemate/constids.inc"
+#include "himbaechel_constids.h"
+
+namespace {
+USING_NEXTPNR_NAMESPACE;
+
+void find_and_bind_downhill_pip(Context *ctx, WireId from, WireId to, NetInfo *net)
+{
+    NPNR_ASSERT(from != WireId());
+    NPNR_ASSERT(to != WireId());
+    for (auto pip : ctx->getPipsDownhill(from)) {
+        if (ctx->getPipDstWire(pip) == to) {
+            log_info("    pip %s: %s -> %s\n", ctx->nameOfPip(pip), ctx->nameOfWire(from), ctx->nameOfWire(to));
```

Comment:
this feels like the kind of thing that should be gated behind `ctx->debug` or `ctx->verbose` ?

Reply:
yeah this is presently incredibly noisy and so most of the debug prints will be gated.